connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java

```diff
@@ -464,8 +465,22 @@ private void convertMessages(ConsumerRecords<byte[], byte[]> msgs) {
         for (ConsumerRecord<byte[], byte[]> msg : msgs) {
             log.trace("{} Consuming and converting message in topic '{}' partition {} at offset {} and timestamp {}",
                     this, msg.topic(), msg.partition(), msg.offset(), msg.timestamp());
-            SchemaAndValue keyAndSchema = keyConverter.toConnectData(msg.topic(), msg.key());
-            SchemaAndValue valueAndSchema = valueConverter.toConnectData(msg.topic(), msg.value());
+            final SchemaAndValue keyAndSchema;
+            final SchemaAndValue valueAndSchema;
+            boolean exceptionCheckForKeyAndSchema = true;
+
+            try {
+                keyAndSchema = keyConverter.toConnectData(msg.topic(), msg.key());
+                exceptionCheckForKeyAndSchema = false;
+                valueAndSchema = valueConverter.toConnectData(msg.topic(), msg.value());
+            } catch (DataException e) {
+                if (exceptionCheckForKeyAndSchema) {
+                    log.error("Failed to convert message Key to Kafka Connect format", e);
+                } else {
+                    log.error("Failed to convert message Value to Kafka Connect format", e);
+                }
+                throw new ConnectException("Exiting WorkerSinkTask due to unconverted message to Kafka Connect format exception", e);
+            }
```

Comment:
Perhaps the following would be a bit more straightforward without the boolean flag while providing more meaningful content in the error message?

            final SchemaAndValue keyAndSchema;
            final SchemaAndValue valueAndSchema;
            try {
                keyAndSchema = keyConverter.toConnectData(msg.topic(), msg.key());
            } catch (Throwable e) {
                String str = String.format("Error converting message key in topic '%s' partition %d at offset %d and timestamp %d",
                                           msg.topic(), msg.partition(), msg.offset(), msg.timestamp());
                throw new ConnectException(str, e);
            }
            try {
                valueAndSchema = valueConverter.toConnectData(msg.topic(), msg.value());
            } catch (Throwable e) {
                String str = String.format("Error converting message value in topic '%s' partition %d at offset %d and timestamp %d",
                                           msg.topic(), msg.partition(), msg.offset(), msg.timestamp());
                throw new ConnectException(str, e);
            }

Also, note that we don't want to log the exception here since exceptions will ultimately be caught and logged in [`WorkerTask.doRun()`](https://github.com/jadireddi/kafka/blob/a967d7b8111f33cd6c95fc5381b274143d6b2319/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerTask.java#L172). Logging it here is unnecessary.

And actually, since there's a fair amount of duplicate logic, a better way of reducing that logic is to extract it into a method:

            SchemaAndValue keyAndSchema = toConnectData(keyConverter, "key", msg, msg.key());
            SchemaAndValue valueAndSchema = toConnectData(valueConverter, "value", msg, msg.value());

with the method:

    private SchemaAndValue toConnectData(Converter converter, String converterName, ConsumerRecord<byte[], byte[]> msg, byte[] data) {
        try {
            return converter.toConnectData(msg.topic(), data);
        } catch (Throwable e) {
            String str = String.format("Error converting message %s in topic '%s' partition %d at offset %d and timestamp %d",
                                       converterName, msg.topic(), msg.partition(), msg.offset(), msg.timestamp());
            throw new ConnectException(str, e);
        }
    }

This is even more concise and readable. WDYT?

Reply:
Thank you @rhauch for your time to review and provide valuable suggestions. Code looks really concise and readable. Making changes.